xdsl/dialects/builtin.py

```diff
@@ -596,6 +596,74 @@ def verify(self, attr: Attribute) -> None:
                                | UnrankedTensorType[_VectorOrTensorElem])
 
 
+@dataclass(init=False)
+class VectorRankConstraint(AttrConstraint):
+    """
+    Constrain a vector to be of a given rank.
+    """
+
+    expected_rank: int
+    """The expected vector rank."""
+
+    def __init__(self, expected_rank: int):
+        self.expected_rank = expected_rank
```

Comment:
This looks like what the generated init would do, maybe remove this and `init=False`?